h/streamer/messages.py

```diff
@@ -103,40 +111,36 @@ def handle_annotation_event(message, sockets, registry, session):
     # Create a generator which has the first socket back again
     matching_sockets = chain((first_socket,), matching_sockets)
 
-    nipsa_service = NipsaService(session)
-    user_nipsad = nipsa_service.is_flagged(annotation.userid)
-
     authority = registry.settings.get("h.authority", "localhost")
-    group_service = GroupfinderService(session, authority)
-    user_service = UserService(authority, session)
-    formatters = [AnnotationUserInfoFormatter(session, user_service)]
+    base_url = registry.settings.get("h.app_url", "http://localhost:5000")
+
+    resource = AnnotationNotificationContext(
+        annotation,
+        group_service=GroupfinderService(session, authority),
+        links_service=LinksService(base_url, registry),
+    )
+    read_principals = principals_allowed_by_permission(resource, "read")
+    reply = _generate_annotation_event(session, authority, message, resource)
```

Comment:
So all this work has been pulled up from `_generate_annotation_event()` and is now done only once, instead of once per socket

Reply:
Pretty much. In some cases we don't do it at all. There were lots of services initialised that are only needed to serialise the annotation. In the event of a delete, we don't need them at all.